### PR TITLE
feat(templates): add user/:id/templates

### PR DESCRIPTION
**Pull request by cheongsiuhong**
Add an endpoint to fetch all templates available to a user (templates which the user is either an Issuer or Editor of).

## Context

Part of #50 
We kind of missed this out.

## Approach

I should have used the query builder :cry:

### CODE REVIEW

File: backend/src/templates/templates.service.ts [hunk 1]
```diff
@@ -138,8 +140,47 @@ export class TemplatesService {
...
+   * @returns All template metadata which user is editor or issuer of.
+   */
+  async listTemplatesForUser(
+    userId: number,
+  ): Promise<GetTemplateMetaResponseDto[]> {
+    const editorOf = await this.connection.manager.find(Editor, {
+      where: {
+        user: userId,
+      },
+      relations: ['template', 'template.author'],
```
Comment (dvrylc) [inline backend/src/templates/templates.service.ts]: Doesn't seem like we need the author relationship?

File: backend/src/templates/templates.service.ts [hunk 2]
```diff
@@ -138,8 +140,47 @@ export class TemplatesService {
...
+    })
 
-  async listTemplatesForUser(): Promise<void> {
-    return
+    const issuerOf = await this.connection.manager.find(Issuer, {
+      where: {
+        user: userId,
+      },
+      relations: ['template', 'template.author'],
```
Comment (dvrylc) [inline backend/src/templates/templates.service.ts]: Same here, author relationship not needed.

File: backend/src/templates/templates.service.ts [hunk 3]
```diff
@@ -138,8 +140,47 @@ export class TemplatesService {
...
+    })
+
+    const templateIds = unionBy(editorOf, issuerOf, (e) => e.template.id).map(
+      (e) => e.template.id,
+    )
+    const templates = await this.connection.manager.find(TemplateVersion, {
+      where: {
+        template: In(templateIds),
```
Comment (dvrylc) [inline backend/src/templates/templates.service.ts]:
```suggestion
        template: { id: In(templateIds) },
```

Although undocumented, it seems like TypeORM matches the primary key by default so this works, but might be a good idea to make it explicit.

File: backend/src/templates/dto/get-template.dto.ts [hunk 4]
```diff
@@ -11,3 +11,13 @@ export class GetTemplateResponseDto {
   body!: TemplateBlock[]
   paramsRequired!: string[]
 }
+
+export class GetTemplateMetaResponseDto {
```
Comment (dvrylc) [inline backend/src/templates/dto/get-template.dto.ts:15]: Sufficient for now, eventually we'll want to differentiate between what the user can do with each template (i.e. editor or issuer).

File: backend/src/users/users.controller.ts [hunk 5]
```diff
@@ -1,19 +1,25 @@
...
   /**
    * List templates that the user has access to
    */
   @Get(':id/templates')
   @ApiResponse({ type: ListTemplatesForUserResponseDto })
-  async listTemplates(@Param('id') _userId: number): Promise<void> {
-    await this.usersService.listTemplates()
+  async listTemplates(
+    @Param('id') _userId: number,
+  ): Promise<GetTemplateMetaResponseDto[]> {
```
Comment (dvrylc) [inline backend/src/users/users.controller.ts]: Incorrect type in `@ApiResponse` decorator

**Review by dvrylc — APPROVED** [review]
Minor comments, can merge once fixed 👍🏻

File: backend/src/templates/dto/get-template.dto.ts (hunk 4 above)
Comment (cheongsiuhong) [reply backend/src/templates/dto/get-template.dto.ts:15]: Fix is up.

**Pull request merged**